lib/block.ml

```diff
@@ -141,165 +119,43 @@ let pp ?syntax ppf b =
-          k (pp_list dump_string) (List.map g ls)
-          (pp_list dump_string) (List.map g ps)
-        :: acc
-    ) [] t.labels
-  |> function
-  | [] -> Result.Ok ()
-  | es -> Result.Error es
-
-let get_label t label =
-  try Some (List.assoc label t.labels)
-  with Not_found -> None
-
-let get_labels t label =
-  List.fold_left (fun acc (k, v) ->
-      if String.equal k label then match v with
-        | None   -> assert false
-        | Some v -> v ::acc
-      else acc
-    ) [] t.labels
-
-let get_prefixed_labels t prefix =
-  List.fold_left (fun acc (k, v) ->
-      if String.equal (String.with_range ~len:(String.length prefix) k) prefix then match v with
-        | None   -> (String.with_range ~first:(String.length prefix) k, None)::acc
-        | Some (e, s) -> (String.with_range ~first:(String.length prefix) k, Some (e, s))::acc
-      else acc
-    ) [] t.labels
-
-let directory t = match get_label t "dir" with
-  | None   -> None
-  | Some None -> None
-  | Some (Some (`Eq, d)) -> Some d
-  | Some (Some _) -> Fmt.failwith "invalid `dir` label value"
-
-let file t = match get_label t "file" with
-  | None   -> None
-  | Some None -> None
-  | Some (Some (`Eq, f)) -> Some f
-  | Some (Some _) -> Fmt.failwith "invalid `file` label value"
-
-let part t = match get_label t "part" with
-  | None   -> None
-  | Some None -> None
-  | Some (Some (`Eq, l)) -> Some l
-  | Some (Some _) -> Fmt.failwith "invalid `part` label value"
-
-let version t = match get_label t "version" with
-  | None -> None
-  | Some None -> None
-  | Some (Some (op, v)) ->
-    match Ocaml_version.of_string v with
-    | Ok v -> Some (op, v)
-    | Error (`Msg e) -> Fmt.failwith "invalid `version` label value: %s" e
+let version t =
+  get_label t (fun acc -> function Version (op, v) -> Some (op, v) | _ -> acc)
 
 let source_trees t =
-  let f = function
-    | `Eq, x -> x
-    | _ -> Fmt.failwith "invalid `source-tree` label value"
-  in
-  List.map f (get_labels t "source-tree")
-
-let mode t = match get_label t "non-deterministic" with
-  | None                  -> `Normal
-  | Some None
-  | Some (Some (`Eq, "output"))  -> `Non_det `Output
-  | Some (Some (`Eq, "command")) -> `Non_det `Command
-  | Some (Some (`Eq, _))         -> `Normal
-  | Some (Some _) -> Fmt.failwith "invalid `non-deterministic` label value"
-
-let skip t = match get_label t "skip" with
-  | Some None -> true
-  | _ -> false
+  List.filter_map (function Label.Source_tree x -> Some x | _ -> None) t.labels
+
+let mode t =
+  List.fold_left
+    (fun acc -> function Label.Non_det mode -> `Non_det mode | _ -> acc)
+    `Normal t.labels
+
+let skip t = List.exists (function Label.Skip -> true | _ -> false) t.labels
 
-let environment t = match get_label t "env" with
-  | None
-  | Some None
-  | Some (Some (`Eq, "default")) -> "default"
-  | Some (Some (`Eq, s)) -> s
-  | Some (Some _) -> Fmt.failwith "invalid `env` label value"
+let environment t =
+  List.fold_left
```

Comment:
How about implementing all those functions on top of `get_label`?

Eventually adding a `get_label_or ~default` variant to be used by mode and env.

Reply:
I improved a bit these functions, but I'm not sure we could use the same base for `get_label` and `get_label_or` without making their usage too complex